pkg/rest/client.go

```diff
@@ -0,0 +1,237 @@
+/*
+ * @license
+ * Copyright 2023 Dynatrace LLC
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package rest
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/log"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/timeutils"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/trafficlogs"
+	"io"
+	"net/http"
+	"time"
+)
+
+type Client struct {
+	client            *http.Client
+	rateLimitStrategy rateLimitStrategy
+	trafficLogger     func(req *http.Request, reqBody string, resp *http.Response, respBody string)
+}
+
+func NewRestClient(client *http.Client, strategy rateLimitStrategy) *Client {
+	return &Client{
+		client:            client,
+		rateLimitStrategy: strategy,
+		trafficLogger:     trafficlogs.LogToFiles,
+	}
+}
+
+func (c Client) Get(ctx context.Context, url string) (Response, error) {
+	req, err := c.request(ctx, http.MethodGet, url)
+
+	if err != nil {
+		return Response{}, err
+	}
+
+	return c.executeRequest(req)
+}
+
+func (c Client) GetWithRetry(ctx context.Context, url string, settings RetrySetting) (resp Response, err error) {
+	resp, err = c.Get(ctx, url)
+
+	if err == nil && resp.IsSuccess() {
+		return resp, nil
+	}
+
+	for i := 0; i < settings.MaxRetries; i++ {
+		log.WithCtxFields(ctx).Warn("Retrying failed GET request %s with error (HTTP %d)", url, resp.StatusCode)
+		time.Sleep(settings.WaitTime)
+		resp, err = c.Get(ctx, url)
+		if err == nil && resp.IsSuccess() {
+			return resp, err
+		}
+	}
+
+	if err != nil {
+		return resp, fmt.Errorf("GET request %s failed after %d retries: %w", url, settings.MaxRetries, err)
+	}
+
+	return resp, RespError{
+		StatusCode: resp.StatusCode,
+		Reason:     fmt.Sprintf("GET request %s failed after %d retries: (HTTP %d)!\n    Response was: %s", url, settings.MaxRetries, resp.StatusCode, resp.Body),
+		Body:       string(resp.Body),
+	}
+}
+
+func (c Client) Delete(ctx context.Context, url string) (Response, error) {
+	req, err := c.request(ctx, http.MethodDelete, url)
+	if err != nil {
```

Comment:
OCD-nitpick: please add a newline before the `if` like in all other methods of the client

Reply:
sure